# Advanced Rate of Return Metrics (IRR, XIRR, MIRR, XMIRR, PV, FV, NPV, XNPV)

## Introduction: The Time Value of Money (TVM)

The **Time Value of Money (TVM)** is a foundational principle in finance. It posits that a sum of money available today is inherently worth more than the same sum promised at a future date. This is due to its potential earning capacity. In essence, a dollar in hand today can be invested and grow, whereas a dollar received tomorrow has missed that opportunity.

**Why is TVM important?**

1.  **Inflation:** Over time, the purchasing power of money tends to erode due to inflation. A fixed sum of money will buy less in the future than it does today.
2.  **Opportunity Cost:** Money available today can be invested or used to generate returns. Delaying the receipt of money means foregoing these potential earnings or opportunities. This foregone return is the opportunity cost.
3.  **Risk/Uncertainty:** There is always a degree of risk that future payments might not be received as expected, or that the entity promising them might default. Money in hand today carries less uncertainty than a future promise.

Understanding TVM is paramount for making financial decisions, from evaluating investment projects and calculating loan payments to personal financial planning. It forms the bedrock for all discounted cash flow (DCF) analyses.

## Building Blocks: Present Value (PV) and Future Value (FV)

To quantify the Time Value of Money, we use two inverse but related concepts: Future Value (FV) and Present Value (PV).

### Future Value (FV)

The **Future Value (FV)** is the value of a current asset at a future date, assuming a specific rate of growth or interest. It answers the question: "What will my money be worth in the future if I invest it today?"

Let's start with a simple question: If you invest $PV$ dollars today at an interest rate of $r$ per period, how much will you have after $n$ periods? This process is called **compounding**.

*   **After 1 period:** You'll have your initial principal plus the interest earned.

    $$FV_1 = PV + (PV \cdot r) = PV \cdot (1+r)$$

*   **After 2 periods:** You'll earn interest on the new, larger amount ($FV_1$). This is the magic of compounding.

    $$FV_2 = FV_1 \cdot (1+r) = [PV \cdot (1+r)] \cdot (1+r) = PV \cdot (1+r)^2$$

*   **After n periods:** Following the pattern, the general formula for Future Value is:

    $$ FV = PV \cdot (1+r)^n $$

**Example:** You invest 1,000 today ($PV$) at an annual interest rate of 8% ($r$) for 5 years ($n$).

$FV = 1000 \cdot (1+0.08)^5 = 1000 \cdot (1.4693) = 1,469.33$

Thus, the final formula is

> $$FV = PV \times (1 + r)^n$$

where:
*   $FV$: Future Value (the amount of money at the end of the investment period)
*   $PV$: Present Value (the initial amount of money invested today)
*   $r$: The annual interest rate, growth rate, or rate of return (expressed as a decimal)
*   $n$: The number of periods (typically years) over which the money is invested

### Present Value (PV) - Discounting

Now let's ask the reverse question: If you are promised $FV$ dollars in $n$ periods, what is the value of that promise *today*, given an interest rate of $r$? This process is called **discounting**, and it's crucial for everything that follows.

We can derive the PV formula simply by rearranging the FV formula. From $FV = PV \cdot (1+r)^n$, we solve for $PV$:

$$ PV = \frac{FV}{(1+r)^n} \quad \text{or} \quad PV = FV \cdot (1+r)^{-n} $$

Here, $r$ is often called the **discount rate**. It's the rate you use to discount future cash flows back to their present value.

**Example:** You will receive 10,000 in 5 years. If your personal discount rate (the return you could get elsewhere) is 8%, what is that $10,000 worth to you today?

$$PV = \frac{10000}{(1+0.08)^5} = \frac{10000}{1.4693} = 6,805.83$$

This means you would be indifferent between receiving $6,805.83 today and receiving $10,000 in 5 years, assuming an 8% rate of return.

The **Present Value (PV)** is the current worth of a future sum of money or a stream of future cash flows, given a specified rate of return.

---

## Net Present Value (NPV)

When evaluating an investment project, we rarely deal with a single future sum but rather a series of cash flows occurring at different times. For example, you might invest money today (an outflow), and then receive a series of returns in the following years (inflows). The **Net Present Value (NPV)** is a powerful and fundamental tool in capital budgeting that uses the concept of Present Value to assess the profitability and attractiveness of such a project.

**Net Present Value (NPV)** is the difference between the present value of all cash inflows and the present value of all cash outflows associated with a project over its entire life.

**Convention:**
*   **Cash Outflows (Investments):** Represented as negative numbers.
*   **Cash Inflows (Returns):** Represented as positive numbers.

The initial investment at "time 0" is denoted as $CF_0$ and is almost always negative.

Essentially, it converts all future cash flows (both positive, representing income, and negative, representing expenses or investments) into their equivalent value today, using a specified discount rate. The initial investment ($CF_0$) is typically a cash outflow and thus negative.

The general formula for NPV is:

$$NPV = \sum_{t=0}^{n} \frac{CF_t}{(1 + r)^t}$$

where:
*   $CF_t$: The net cash flow (cash inflow minus cash outflow) at time $t$.
    *   $CF_0$ represents the initial investment at time $t=0$, which is almost always an outflow and thus a negative value.
*   $r$: The discount rate (also known as the hurdle rate, required rate of return, or cost of capital). This rate is of paramount importance as it represents the minimum acceptable rate of return for the project, often reflecting the company's cost of financing (e.g., weighted average cost of capital), the risk profile of the project, and the opportunity cost of investing elsewhere.
*   $t$: The time period (e.g., year) when the cash flow occurs. The first cash flow is typically at $t=0$.
*   $n$: The total number of periods over the project's life.

---

In simple terms, the cost of capital is the minimum return a company must earn on a project or investment to make it worthwhile. It's essentially the cost of the money a company uses to fund its operations, whether that money comes from borrowing or from investors.

Think of it like this: if you borrow money from a bank to start a business, the interest you pay on that loan is a cost. Your business needs to make at least enough profit to cover that interest. The cost of capital is a similar idea, but it also includes the "cost" of money from investors who expect a return on their investment.

Companies use the cost of capital as a benchmark to decide if a new project is a good idea. If the expected return from the project is higher than the cost of capital, the project is likely to be profitable. If it's lower, the company will probably lose money on it.

**The Two Main Ingredients: Debt and Equity**

A company's capital usually comes from two main sources: debt and equity.

**1. Cost of Debt:** This is the easier one to understand. It's the interest a company pays on its loans. For example, if a company takes out a $100,000 loan with a 5% interest rate, its cost of debt is 5%. Because interest payments are usually tax-deductible, the actual cost of debt is often lower than the interest rate.

**2. Cost of Equity:** This is the return that shareholders expect for investing in a company. Unlike the fixed interest payments on a loan, there's no set payment for equity investors. However, they still expect to be compensated for the risk they are taking by investing their money in the company. This return comes in the form of dividends and an increase in the stock's value. The cost of equity is generally higher than the cost of debt because shareholders are paid last if a company goes bankrupt.

**Putting It All Together: The Weighted Average Cost of Capital (WACC)**

Most companies use a mix of both debt and equity to finance their operations. To find the overall cost of capital, they calculate the **Weighted Average Cost of Capital (WACC)**. This calculation takes into account the proportion of debt and equity a company uses and their respective costs.

The formula might look complicated, but the concept is straightforward: it's the average cost of all the capital a company has, weighted by how much of each type of capital it uses.

**A Simple Example**

Let's say a company, "Creative Innovations Inc.," is funded in the following way:
*   **60% from equity** (investors) who expect a **10%** return.
*   **40% from debt** (a bank loan) with an after-tax interest rate of **4%**.

Here's how to calculate the WACC for Creative Innovations Inc.:

1.  **Multiply the weight of equity by its cost:** 60% * 10% = 6.0%
2.  **Multiply the weight of debt by its cost:** 40% * 4% = 1.6%
3.  **Add the two together:** 6.0% + 1.6% = **7.6%**

So, the WACC for Creative Innovations Inc. is **7.6%**.

This means that for any new project the company considers, it must expect a return of at least 7.6% to be profitable and create value for its investors. If a project is expected to return 12%, it's a great investment. If it's only expected to return 5%, the company would be better off not pursuing it.

---

**Explanation of the summation:**

The formula $NPV = \sum_{t=0}^{n} \frac{CF_t}{(1 + r)^t}$ means you calculate the present value of each individual cash flow and then sum them up:

$$NPV = \frac{CF_0}{(1 + r)^0} + \frac{CF_1}{(1 + r)^1} + \frac{CF_2}{(1 + r)^2} + \dots + \frac{CF_n}{(1 + r)^n}$$

Since any non-zero number raised to the power of $0$ is $1$, the first term simplifies to $CF_0$.

**Interpretation**

The decision rule for NPV is straightforward and unambiguous:
*   If $NPV > 0$: The project is expected to generate more cash than the cost of capital, after accounting for the time value of money. This means the project will increase the wealth of the shareholders. **Accept** the project.
*   If $NPV < 0$: The project is expected to generate less cash than the cost of capital, indicating it will destroy shareholder wealth. **Reject** the project.
*   If $NPV = 0$: The project is expected to break even, covering the cost of capital exactly. The investor is indifferent between undertaking the project and not.

### Examples

**Example 1:** A project requires an initial investment of $10.000$ today ($CF_0 = -10,000$). It is expected to generate cash inflows of $4,000$ at the end of Year 1 ($CF_1$), $5,000$ at the end of Year 2 ($CF_2$), and $6,000$ at the end of Year 3 ($CF_3$). Assume a discount rate ($r$) of $10\%$.

$NPV = \frac{-10,000}{(1 + 0.10)^0} + \frac{4,000}{(1 + 0.10)^1} + \frac{5,000}{(1 + 0.10)^2} + \frac{6,000}{(1 + 0.10)^3}$

$NPV = -10,000 + \frac{4,000}{1.10} + \frac{5,000}{1.21} + \frac{6,000}{1.331}$

$NPV = -10,000 + 3,636.36 + 4,132.23 + 4,507.89$

$NPV = 2276.48$

Since $NPV > 0$, this project would be accepted.

In [1]:
import numpy_financial as npf

cash_flows = [-10000, 4000, 5000, 6000]

discount_rate = 0.10  # 10%

# The npv function calculates the NPV of cash flows from year 1 onwards
# So we need to calculate it and then add the initial investment (CF0)
# Note: npf.npv(rate, values) takes the rate and then a list of values from CF1, CF2...
npv_at_discount_rate = npf.npv(discount_rate, cash_flows)

print(f"The NPV of the project at its discount rate of {discount_rate:.2%} is: ${npv_at_discount_rate:.4f}")

The NPV of the project at its discount rate of 10.00% is: $2276.4838


In [2]:
from pyxirr import npv

cash_flows = [-10000, 4000, 5000, 6000]

discount_rate = 0.10  # 10%

# pyxirr's npv function includes all cash flows (CF0, CF1, CF2, ...)
# So we pass the entire cash_flows list directly
npv_at_discount_rate = npv(discount_rate, cash_flows)

print(f"The NPV of the project at its discount rate of {discount_rate:.2%} is: ${npv_at_discount_rate:.4f}")

The NPV of the project at its discount rate of 10.00% is: $2276.4838


**Example 2:** A project requires an initial investment of $10.000$ today ($CF_0 = -10,000$). It is expected to generate cash inflows of $4,000$ at the end of Year 1 ($CF_1$), $5,000$ at the end of Year 2 ($CF_2$), and $6,000$ at the end of Year 3 ($CF_3$). Assume a discount rate ($r$) of $25\%$.

$NPV = \frac{-10,000}{(1 + 0.25)^0} + \frac{4,000}{(1 + 0.25)^1} + \frac{5,000}{(1 + 0.25)^2} + \frac{6,000}{(1 + 0.25)^3}$

$NPV = -10,000 + \frac{4,000}{1.25} + \frac{5,000}{1.5625} + \frac{6,000}{1.953125}$

$NPV = -10,000 + 3,200.00 + 3,200.00 + 3,072.00$

$NPV = -528.00$

Since the NPV is negative, this project is not a good investment at a 25% discount rate. It fails to generate enough return to justify the cost.

In [3]:
import numpy_financial as npf

cash_flows = [-10000, 4000, 5000, 6000]

discount_rate = 0.25  # 25%

# The npv function calculates the NPV of cash flows from year 1 onwards
# So we need to calculate it and then add the initial investment (CF0)
# Note: npf.npv(rate, values) takes the rate and then a list of values from CF1, CF2...
npv_at_discount_rate = npf.npv(discount_rate, cash_flows)

print(f"The NPV of the project at its discount rate of {discount_rate:.2%} is: ${npv_at_discount_rate:.4f}")

The NPV of the project at its discount rate of 25.00% is: $-528.0000


In [4]:
from pyxirr import npv

cash_flows = [-10000, 4000, 5000, 6000]

discount_rate = 0.25  # 25%

# pyxirr's npv function includes all cash flows (CF0, CF1, CF2, ...)
# So we pass the entire cash_flows list directly
npv_at_discount_rate = npv(discount_rate, cash_flows)

print(f"The NPV of the project at its discount rate of {discount_rate:.2%} is: ${npv_at_discount_rate:.4f}")

The NPV of the project at its discount rate of 25.00% is: $-528.0000


**Example 3:** Consider a project with a required rate of return (discount rate) of 10% ($r=0.10$).
*   Initial Investment ($CF_0$): -$100,000
*   Year 1 Inflow ($CF_1$): +$30,000
*   Year 2 Inflow ($CF_2$): +$40,000
*   Year 3 Inflow ($CF_3$): +$50,000

$NPV = -100000 + \frac{30000}{(1+0.10)^1} + \frac{40000}{(1+0.10)^2} + \frac{50000}{(1+0.10)^3}$

$NPV = -100000 + \frac{30000}{1.10} + \frac{40000}{1.21} + \frac{50000}{1.331}$

$NPV = -100000 + 27272.73 + 33057.85 + 37565.74$

$NPV = -2103.68$

Since the NPV is negative, this project is not a good investment at a 10% discount rate. It fails to generate enough return to justify the cost.

---

In [5]:
import numpy_financial as npf

cash_flows = [-100000, 30000, 40000, 50000]

discount_rate = 0.1  # 10%

# The npv function calculates the NPV of cash flows from year 1 onwards
# So we need to calculate it and then add the initial investment (CF0)
# Note: npf.npv(rate, values) takes the rate and then a list of values from CF1, CF2...
npv_at_discount_rate = npf.npv(discount_rate, cash_flows)

print(f"The NPV of the project at its discount rate of {discount_rate:.2%} is: ${npv_at_discount_rate:.4f}")

The NPV of the project at its discount rate of 10.00% is: $-2103.6814


In [6]:
from pyxirr import npv

cash_flows = [-100000, 30000, 40000, 50000]

discount_rate = 0.1  # 10%

# pyxirr's npv function includes all cash flows (CF0, CF1, CF2, ...)
# So we pass the entire cash_flows list directly
npv_at_discount_rate = npv(discount_rate, cash_flows)

print(f"The NPV of the project at its discount rate of {discount_rate:.2%} is: ${npv_at_discount_rate:.4f}")

The NPV of the project at its discount rate of 10.00% is: $-2103.6814


## Internal Rate of Return (IRR)

The **Internal Rate of Return (IRR)** is another widely used metric for evaluating the attractiveness of a project or investment. While related to NPV, it provides a different perspective on profitability.

NPV answers the question: "Is this a good investment at a *given* discount rate?".

The **Internal Rate of Return (IRR)** answers a different, but related, question: **"What is the discount rate at which this investment breaks even?"**

The **IRR** is the discount rate that makes the Net Present Value (NPV) of all cash flows from a particular project exactly equal to zero. In simpler terms, it is the rate of return an investment is expected to generate. It's the maximum effective interest rate an investment can have before it starts losing money (i.e., having a negative NPV).

**Formula (Implicit)**

Unlike NPV, there is no direct algebraic formula to explicitly calculate IRR. Instead, IRR is the specific value of $IRR$ that satisfies the following equation:

$$0 = \sum_{t=0}^{n} \frac{CF_t}{(1 + IRR)^t}$$

$$ 0 = CF_0 + \frac{CF_1}{(1+r_{IRR})^1} + \frac{CF_2}{(1+r_{IRR})^2} + \dots + \frac{CF_n}{(1+r_{IRR})^n} $$

This is essentially the NPV formula, but with the NPV set to zero and $IRR$ as the unknown discount rate we are solving for.

Because $IRR$ appears in the denominator and as an exponent, this equation is not algebraically solvable for most projects (especially those with more than two cash flows). We need to employ iterative numerical methods (such as the Newton-Raphson method, which we will discuss for XIRR) to efficiently find the IRR.

**Interpretation**

The decision rule for IRR is:
*   If $IRR > r$: Where $r$ is the required rate of return (hurdle rate or cost of capital), the project's expected return exceeds the minimum acceptable return. **Accept** the project.
*   If $IRR < r$: The project's expected return is less than the cost of capital. **Reject** the project.
*   If $IRR = r$: The project's expected return exactly equals the cost of capital. The investor is indifferent.

**Example (continued from NPV example):** For the project with $CF_0 = -10,000$, $CF_1 = 4,000$, $CF_2 = 5,000$, and $CF_3 = 6,000$. We need to find the $IRR$ such that:

$0 = \frac{-10,000}{(1 + IRR)^0} + \frac{4,000}{(1 + IRR)^1} + \frac{5,000}{(1 + IRR)^2} + \frac{6,000}{(1 + IRR)^3}$

Using `numpy_financial` library (see example below), the $IRR$ for this project is approximately $21.65\%$.
If our required rate of return ($r$) is $10\%$, then since $21.65\% > 10\%$, we would accept the project.

In [7]:
import numpy_financial as npf

# Define the cash flows
# CF0 = -10,000 (initial investment)
# CF1 = +4,000
# CF2 = +5,000
# CF3 = +6,000
cash_flows = [-10000, 4000, 5000, 6000]

# Calculate the Internal Rate of Return (IRR)
irr_value = npf.irr(cash_flows)

# Print the result in a readable format
print(f"The project's cash flows are: {cash_flows}")
print(f"The calculated IRR is: {irr_value:.4f}")
print(f"As a percentage, the IRR is: {irr_value:.2%}")

The project's cash flows are: [-10000, 4000, 5000, 6000]
The calculated IRR is: 0.2165
As a percentage, the IRR is: 21.65%


In [8]:
from pyxirr import irr

# Define the cash flows
# CF0 = -10,000 (initial investment)
# CF1 = +4,000
# CF2 = +5,000
# CF3 = +6,000
cash_flows = [-10000, 4000, 5000, 6000]

# Calculate the Internal Rate of Return (IRR)
irr_value = irr(cash_flows)

# Print the result in a readable format
print(f"The project's cash flows are: {cash_flows}")
print(f"The calculated IRR is: {irr_value:.4f}")
print(f"As a percentage, the IRR is: {irr_value:.2%}")

The project's cash flows are: [-10000, 4000, 5000, 6000]
The calculated IRR is: 0.2165
As a percentage, the IRR is: 21.65%


**Limitations of IRR**

While intuitive, IRR has several limitations that can make it less reliable than NPV in certain situations:

1.  **Assumes Regular Intervals:** The standard IRR formula implicitly assumes that all cash flows occur at equally spaced intervals (e.g., annually, semi-annually). This is often an unrealistic simplification in real-world projects. What if you invest on Jan 5th, get a dividend on March 20th, invest more on Sep 1st, and sell everything two years later on Dec 18th? For this, we need a more flexible tool.
2.  **Reinvestment Assumption:** IRR implicitly assumes that all intermediate positive cash flows generated by the project are reinvested at the *calculated IRR itself* until the end of the project's life. This can be a highly unrealistic assumption, especially if the calculated IRR is very high or very low, as it's unlikely that a firm can consistently find other projects with the exact same rate of return to reinvest these funds.
3.  **Multiple IRRs:** For projects with "non-conventional" cash flow patterns (i.e., cash flows that switch signs more than once, e.g., an initial outflow, followed by inflows, then another outflow), it's mathematically possible to have multiple real IRRs, making the interpretation ambiguous and the decision rule unreliable.
4.  **Mutually Exclusive Projects:** When comparing mutually exclusive projects (where choosing one automatically precludes choosing the others), IRR can sometimes lead to different decisions than NPV, especially when projects differ significantly in scale, project life, or the timing of cash flows. In such cases, NPV is generally preferred because it directly measures the absolute increase in firm value.

## NPV vs. IRR: A Comparative Perspective

Both NPV and IRR are powerful tools for capital budgeting, and for independent projects (where accepting one doesn't affect the decision to accept another), they generally lead to the same accept/reject decision. However, conflicts can arise, particularly with mutually exclusive projects, and it's crucial to understand their differences.

| Feature             | Net Present Value (NPV)                                    | Internal Rate of Return (IRR)                                    |
| :------------------ | :--------------------------------------------------------- | :--------------------------------------------------------------- |
| **Output**          | A dollar amount (absolute value)                           | A percentage (rate of return)                                    |
| **Reinvestment**    | Assumes cash flows are reinvested at the specified discount rate ($r$, typically the cost of capital). This is generally a more realistic assumption. | Assumes cash flows are reinvested at the calculated IRR itself. This can be an unrealistic assumption. |
| **Consistency**     | Always consistent with the goal of maximizing shareholder wealth.                                                          | Can conflict with NPV for mutually exclusive projects or non-conventional cash flows. |
| **Multiple Rates**  | Always a unique NPV for a given discount rate.             | Can have multiple IRRs with non-conventional cash flows, leading to ambiguity. |
| **Decision Rule**   | $NPV > 0 \implies$ Accept                                    | $IRR > r \implies$ Accept                                        |
| **Primary Use**     | Most reliable and theoretically sound for capital budgeting decisions, especially for comparing and ranking mutually exclusive projects. | Intuitive for managers (as a percentage return); good for communicating project returns when no conflicts arise. |

**Why NPV is generally preferred for mutually exclusive projects:**

When choosing among mutually exclusive projects, the goal is to select the project that adds the most value to the firm. NPV directly measures this absolute increase in wealth. A project with a higher IRR might not necessarily have the highest NPV if it's a smaller project or if its cash flows are front-loaded, leading to a higher percentage return on a smaller base.

For example, consider two projects with a cost of capital of **10%**:
*   **Project A:** Initial investment $-1,000$, Year 1 cash flow $1,500$. IRR = 50%, and $NPV = 363.64$ (calculation is 1500/(1+0.1) - 1000).
*   **Project B:** Initial investment $-10,000$, Year 1 cash flow $12,000$. IRR = 20%, and $NPV = 909.09$ (calculation is 12000/(1+0.1) - 10000).

Project A has a much higher IRR, but Project B has a significantly higher NPV. If these are mutually exclusive, Project B should be chosen because it adds more absolute value to the firm, despite its lower percentage return. The discrepancy often arises from the differing reinvestment rate assumptions (IRR assumes reinvestment at the project's own rate, while NPV assumes reinvestment at the cost of capital). Therefore, for choosing among alternatives that cannot all be undertaken, NPV is the superior metric.


## Addressing Limitations: Extended Internal Rate of Return (XIRR)

The primary limitation that the **Extended Internal Rate of Return (XIRR)** addresses is the assumption of regularly spaced cash flows inherent in the standard IRR calculation.

Many real-world investments and projects do not have cash flows that occur at precise, equally spaced intervals (e.g., exactly annually or semi-annually). Examples include:
*   Irregular contributions or withdrawals from an investment portfolio.
*   Project expenses or revenues that occur on specific, non-annual dates.
*   Private equity investments with multiple capital calls and distributions over varying timeframes.

Standard IRR would require approximations or complex adjustments to handle such irregular timings, leading to inaccuracies. XIRR provides a robust solution by incorporating the **exact dates** of each cash flow.

The **Extended Internal Rate of Return (XIRR)** is a financial metric used to calculate the annualized internal rate of return for a series of cash flows that are not necessarily periodic. It accounts for the actual, specific dates of each cash flow, providing a more accurate annualized rate of return for investments with irregular timings. Like IRR, XIRR is the discount rate that makes the Net Present Value (NPV) of a series of cash flows equal to zero.

The XIRR is the value of $r$ that satisfies the equation:

$$0 = \sum_{i=1}^{N} \frac{C_i}{(1 + r)^{(D_i - D_1)/365}}$$

where:
*   $C_i$: The $i$-th cash flow amount. This series must include at least one negative cash flow (typically an initial investment or contribution) and at least one positive cash flow (a return or distribution).
*   $D_i$: The date of the $i$-th cash flow. These dates must be actual calendar dates.
*   $D_1$: The date of the **earliest** cash flow in the series. This date serves as the reference point (effectively "time $t=0$") for all other cash flows in the calculation.
*   $(D_i - D_1)$: Represents the number of days between the $i$-th cash flow date and the first (earliest) cash flow date.
*   $/365$: Divides the number of days by $365$ to convert the time difference into a fractional number of years. This uses a common **day count convention** known as "Actual/365 Fixed" (or "Actual/365"). In this convention, the actual number of days between two dates is used in the numerator, and a fixed denominator of 365 days is used for annualization, regardless of whether it's a leap year. This simplification ensures consistency across calculations, though other day count conventions (like Actual/Actual or 30/360) are used in specific financial markets (e.g., bond pricing). Some models use 365.25 for leap years, but 365 is standard in most spreadsheet functions.
*   $N$: The total number of cash flows in the series.

**Important Note on Indexing:** When applying this formula, the first cash flow $C_1$ (corresponding to date $D_1$) will have an exponent of $(D_1 - D_1)/365 = 0$. Consequently, its term in the summation will be $C_1 / (1+r)^0 = C_1$. All subsequent cash flows are then discounted back to this initial reference date $D_1$.

**How XIRR is Solved (Numerical Method)**

Similar to the standard IRR, XIRR cannot be solved directly with a simple algebraic formula. This is because the rate $r$ appears in the exponent of multiple terms, and these exponents are now fractional and often irregular. Instead, XIRR is solved using **iterative numerical methods**.

A commonly employed method is the **Newton-Raphson method**. This algorithm works by iteratively refining an initial guess for $r$ until the NPV of the cash flows (calculated using the XIRR formula) is sufficiently close to zero.

Let $f(r) = \sum_{i=1}^{N} \frac{C_i}{(1 + r)^{(D_i - D_1)/365}}$. Our objective is to find $r$ such that $f(r) = 0$.

The Newton-Raphson iteration formula is:

$$r_{new} = r_{old} - \frac{f(r_{old})}{f'(r_{old})}$$

Where $f'(r)$ is the first derivative of $f(r)$ with respect to $r$:

$$f'(r) = \sum_{i=1}^{N} C_i \times \left( -\frac{D_i - D_1}{365} \right) \times (1 + r)^{-\left(\frac{D_i - D_1}{365} + 1\right)}$$

The algorithm starts with an initial guess ($r_{old}$), calculates a new estimate ($r_{new}$), and repeats this process. The iteration continues until the absolute difference between $r_{new}$ and $r_{old}$ becomes very small, or until the value of $f(r_{new})$ is extremely close to zero, indicating convergence to a solution.


## XIRR Function in Google Sheets

Google Sheets provides a convenient built-in function to calculate XIRR, abstracting away the complex numerical iteration process.

The syntax for the XIRR function in Google Sheets is:

`=XIRR(cash_flow_amounts, cash_flow_dates, [guess])`

Arguments:
*   `cash_flow_amounts`: This is a required argument. It is a range or array containing the numerical values of the cash flows.
    *   It must include at least one negative value (representing an investment, outflow, or cost) and at least one positive value (representing a return, inflow, or revenue).
    *   The order of cash flow amounts must directly correspond to the order of their respective dates.
*   `cash_flow_dates`: This is also a required argument. It is a range or array containing the dates corresponding to each cash flow amount.
    *   These dates must be valid Google Sheets date formats (e.g., entered as `DATE(YYYY,MM,DD)`, `MM/DD/YYYY`, or recognized date serial numbers).
    *   For clarity and to prevent potential internal sorting issues, it is best practice to provide the dates in chronological order, although the function often handles unsorted dates by implicitly sorting them.
*   `[guess]`: This is an optional argument. It is an estimate for the XIRR, expressed as a decimal (e.g., $0.1$ for $10\%$).
    *   If omitted, Google Sheets uses a default guess (typically $0.1$ or $10\%$).
    *   Providing a reasonable guess can be critically important. It helps the iterative numerical algorithm converge faster and, more importantly, helps to find the most economically meaningful solution, especially in cases of unusual or complex cash flow patterns where multiple mathematical solutions (IRRs) might exist. If the function returns an error (e.g., `#NUM!`) or an unexpected result, trying a different `guess` is often the first troubleshooting step.

The `XIRR` function in Google Sheets calculates the **annualized effective yield** (or rate of return) of an investment that has an irregular series of cash flows. It achieves this by finding the discount rate that sets the Net Present Value of these irregularly timed cash flows to zero. The output is always an annualized percentage.

### Example in Google Sheets

Let's assume you have the following cash flows and dates in your Google Sheet:

| Date          | Cash Flow Amount |
| :------------ | :--------------- |
| `2022-01-15`  | `-10000`         |
| `2022-06-01`  | `-2000`          |
| `2023-03-10`  | `5000`           |
| `2024-07-22`  | `10000`          |
| `2024-12-31`  | `3000`           |

If these dates are in cells `A2:A6` and the amounts are in `B2:B6`, you would use the formula:

`=XIRR(B2:B6, A2:A6)`

The result would be approximately **0.2124**.

**Interpreting the Result:** If you format this cell as a percentage, it will display **21.24%**. This means that this specific sequence of investments and returns is equivalent to putting your money in a bank account that pays a **21.24% annualized interest rate**. It's a powerful and accurate way to measure the true performance of your investments.

If you wanted to provide a guess, say $15\%$ (or $0.15$), you could use: `=XIRR(B2:B6, A2:A6, 0.15)`

Here is a Python program which calculates the same example:

In [9]:
from pyxirr import xirr
from datetime import date

# Define the dates and cash flows
dates = [
    date(2022, 1, 15),
    date(2022, 6, 1),
    date(2023, 3, 10),
    date(2024, 7, 22),
    date(2024, 12, 31)
]

cash_flows = [-10000, -2000, 5000, 10000, 3000]

# Calculate the XIRR (Internal Rate of Return for irregular intervals)
xirr_value = xirr(dates, cash_flows)

# Print the results
print("Date\t\t\tCash Flow")
print("-" * 40)
for d, cf in zip(dates, cash_flows):
    print(f"{d}\t\t{cf:>8}")
print("-" * 40)
print(f"\nThe calculated XIRR is: {xirr_value:.4f}")
print(f"As a percentage, the XIRR is: {xirr_value:.2%}")

Date			Cash Flow
----------------------------------------
2022-01-15		  -10000
2022-06-01		   -2000
2023-03-10		    5000
2024-07-22		   10000
2024-12-31		    3000
----------------------------------------

The calculated XIRR is: 0.2124
As a percentage, the XIRR is: 21.24%


## Important Considerations for XIRR

1.  **Annualized Rate:** XIRR always returns an annualized rate, regardless of the frequency of cash flows or the total duration of the investment. This is a key feature, allowing for direct comparison with other annualized investment returns (e.g., bond yields, stock market returns).
2.  **Reinvestment Assumption:** Like standard IRR, XIRR implicitly assumes that all positive cash flows generated by the project are reinvested at the *calculated XIRR rate* until the end of the project. This can be a strong and often unrealistic assumption, as finding consistent reinvestment opportunities at the project's exact rate of return can be challenging.
    *   **Addressing the Reinvestment Assumption: Modified Internal Rate of Return (MIRR):** To overcome the unrealistic reinvestment assumption of IRR/XIRR, the **Modified Internal Rate of Return (MIRR)** was developed. MIRR assumes that all positive cash flows are reinvested at a more realistic rate, typically the firm's cost of capital or a specific finance rate. All negative cash flows are discounted to the present at the firm's finance rate, while positive cash flows are compounded to the project's end at the reinvestment rate, resulting in a terminal value. This terminal value is then discounted back to the present at the cost of capital to find a single, more realistic rate of return.
3.  **Data Quality:** The accuracy and reliability of the XIRR calculation are critically dependent on the accuracy of both the cash flow amounts and, especially, their corresponding dates. Even minor errors in dates can significantly distort the resulting XIRR.
4.  **At Least One Negative and One Positive Cash Flow:** The XIRR function requires a mix of cash flow signs (at least one negative and at least one positive) to find a meaningful rate. If all cash flows are positive (e.g., just receiving money without an initial investment) or all are negative (e.g., only making investments without any return), XIRR will typically return an error (e.g., `#NUM!` in Google Sheets) or an undefined result, as a rate of return cannot be calculated in such scenarios.
5.  **Chronological Order of Dates:** While spreadsheet functions like Google Sheets' XIRR might internally sort the provided dates, it is always best practice to arrange your cash flow dates and corresponding amounts in strict chronological order. This improves readability, ensures clarity, and helps prevent potential unexpected behavior or errors in complex datasets.
6.  **Multiple Solutions / No Real Solution:** For unusual cash flow patterns that involve multiple sign changes (e.g., initial outflow, then inflows, then another outflow, and then more inflows), it is theoretically possible for the underlying XIRR equation to have multiple real solutions or, in some cases, no real solution. This is where the optional `[guess]` argument becomes particularly important. By providing a reasonable `guess`, you can help the numerical algorithm converge to the most appropriate or economically meaningful solution, or even find a solution when the default guess might fail.
7.  **Day Count Convention (Actual/365 Fixed):** As discussed, the XIRR formula's use of `/365` implies the "Actual/365 Fixed" day count convention. This means the numerator uses the exact number of days between two dates, while the denominator is a fixed 365 days for annualization, ignoring leap years. This convention is common in many financial calculations for simplicity and consistency.

---

## Extended Net Present Value (XNPV)

Just as XIRR addresses the limitation of fixed time periods in the standard IRR calculation, the **Extended Net Present Value (XNPV)** provides a more precise valuation when cash flows occur at irregular intervals.

While the standard NPV formula is robust, it assumes that each cash flow ($CF_t$) occurs at the end of a regular, equal-length period (e.g., exactly one year, two years, etc., from the start). In reality, project cash flows often don't adhere to such a strict schedule. An investment might be made mid-month, with returns occurring on various dates throughout the subsequent years. Using the standard NPV formula in such cases requires approximating the timing, which can lead to inaccuracies.

The **Extended Net Present Value (XNPV)** is a financial metric that calculates the net present value of a series of cash flows occurring at specific, irregular dates. By using the actual date of each cash flow, XNPV provides a more accurate valuation than the standard NPV function, which assumes periodic cash flows.

The XNPV formula discounts each individual cash flow from its specific date back to the start date of the investment. It is an application of the fundamental discounting principle, now applied over a more precise time frame measured in days rather than whole periods.

The formula for XNPV is:

$$XNPV = \sum_{i=1}^{N} \frac{C_i}{(1 + r)^{(D_i - D_1)/365}}$$

where:
*   **$C_i$**: The i-th cash flow amount.
*   **$D_i$**: The date of the i-th cash flow.
*   **$D_1$**: The date of the first cash flow, which serves as the initial reference point.
*   **$r$**: The annual discount rate.
*   **$(D_i - D_1)/365$**: This term calculates the exact number of years (including fractions of a year) between each cash flow and the very first cash flow. This precise time measurement is what distinguishes XNPV from NPV.

The interpretation of the XNPV result is identical to the standard NPV:
*   **If XNPV > 0**: The investment is expected to be profitable and add value. **Accept** the project.
*   **If XNPV < 0**: The investment is expected to be unprofitable and destroy value. **Reject** the project.
*   **If XNPV = 0**: The investment is expected to break even.

### Example of XNPV

Let's adapt the XIRR example to calculate the XNPV. Assume your required rate of return (discount rate) is 10%.

The cash flows are as follows:

| Date          | Cash Flow Amount |
| :------------ | :--------------- |
| `2022-01-15`  | `-10000`         |
| `2022-06-01`  | `-2000`          |
| `2023-03-10`  | `5000`           |
| `2024-07-22`  | `10000`          |
| `2024-12-31`  | `3000`           |

**Calculation Breakdown:**

*   **Discount Rate (r):** 10% or 0.10
*   **First Date (D1):** Jan 15, 2022


In [10]:
from pyxirr import xnpv
from datetime import date

# Define the dates and cash flows
dates = [
    date(2022, 1, 15),
    date(2022, 6, 1),
    date(2023, 3, 10),
    date(2024, 7, 22),
    date(2024, 12, 31)
]

cash_flows = [-10000, -2000, 5000, 10000, 3000]

# Define the discount rate
discount_rate = 0.10  # 10%

# Calculate the XNPV
xnpv_value = xnpv(discount_rate, dates, cash_flows)

print(f"The calculated XNPV at a {discount_rate:.2%} discount rate is: ${xnpv_value:.2f}")

The calculated XNPV at a 10.00% discount rate is: $2680.79


The result of this calculation is **$2,680.79**. Since the XNPV is positive, this investment would be considered financially attractive at a 10% discount rate.

---

## Addressing Limitations: Modified Internal Rate of Return (MIRR)

As noted earlier, one of the most significant drawbacks of the standard IRR (and XIRR) is its **reinvestment assumption**. It implicitly assumes that all positive cash flows generated by a project are reinvested at the IRR itself. This can be unrealistic, especially if the IRR is very high. It's unlikely a company can consistently find new investment opportunities that yield the exact same high rate of return.

The **Modified Internal Rate of Return (MIRR)** was developed to address this issue by providing a more realistic measure of a project's profitability. MIRR allows for the specification of a different, more practical reinvestment rate for the positive cash flows.

**Key Differences from IRR:**
*   **IRR Assumption:** Intermediate cash flows are reinvested at the IRR.
*   **MIRR Assumption:** Intermediate cash flows are reinvested at a rate that reflects the firm's actual investment opportunities, typically the cost of capital.

### Derivation and Formula

The calculation of MIRR involves three main steps:

1.  **Calculate the Present Value (PV) of all negative cash flows (outflows).** All investment costs are discounted to their value at time t=0 using a **finance rate**. This rate is typically the firm's cost of capital or the specific cost of borrowing funds for the project.
2.  **Calculate the Future Value (FV) (not Present value) of all positive cash flows (inflows).** All returns are compounded forward to the end of the project's life (time t=n) using a **reinvestment rate**. This rate should reflect the return the company expects to earn on reinvesting these funds (e.g., the weighted average cost of capital). This single compounded sum is known as the **Terminal Value**.
3.  **Calculate the MIRR.** The MIRR is the discount rate that equates the present value of the outflows with the present value of the Terminal Value. In other words, it's the rate of return that makes the future value of the initial investments grow to the Terminal Value over the project's life.

The formula for MIRR is expressed as:

$$ MIRR = \left( \frac{\text{FV of Positive Cash Flows (at Reinvestment Rate)}}{\text{PV of Negative Cash Flows (at finance rate)}} \right)^{\frac{1}{n}} - 1 $$

where $n$ is the number of periods.

This formula is derived from the following:

$$ \text{PV of Negative Cash Flows (at finance rate)} = \frac{\text{FV of Positive Cash Flows (at Reinvestment Rate)}}{(1 + MIRR)^n}$$


### Example of MIRR

Let's use our familiar cash flow series: an initial investment of 10,000 followed by inflows of 4,000, 5,000, and 6,000 over three years.

*   **Cash Flows:** `[-10000, 4000, 5000, 6000]`
*   Assume the **finance rate** (cost to borrow the initial 10,000) is 10%.
*   Assume the **Reinvestment Rate** (the rate at which you can reinvest the inflows) is 12%.

**Step-by-Step Calculation:**

1.  **Present Value of Negative Cash Flows:**

    There is only one negative cash flow, the initial investment of 10,000 at t=0. Its present value is simply **10,000**.

2.  **Future Value of Positive Cash Flows (Terminal Value):**

    We need to compound each inflow to the end of Year 3 at the 12% reinvestment rate.
    *   The 4,000 from Year 1 will be reinvested for 2 years: $FV = 4,000 \times (1 + 0.12)^2 = 4,000 \times 1.2544 = 5,017.60$
    *   The 5,000 from Year 2 will be reinvested for 1 year: $FV = 5,000 \times (1 + 0.12)^1 = 5,000 \times 1.12 = 5,600.00$
    *   The 6,000 from Year 3 is already at the end, so its future value is **6,000.00**.
    *   **Terminal Value** = 5,017.60 + 5,600.00 + 6,000.00 = **16,617.60**

3.  **Calculate the MIRR:**

    Now we find the rate that makes the initial $10,000 investment grow to the 16,617.60 Terminal Value in 3 years.

    $$ MIRR = \left( \frac{16617.60}{10000} \right)^{\frac{1}{3}} - 1 $$
    $$ MIRR = (1.66176)^{\frac{1}{3}} - 1 $$
    $$ MIRR = 1.1845 - 1 = 0.1845 $$

The MIRR is **18.45%**.

Notice this is lower than the previously calculated IRR of **21.65%**. This is because the MIRR uses a more conservative and realistic reinvestment rate (12%) compared to the IRR's implicit assumption of reinvesting at 21.65%. This makes MIRR a more trustworthy metric for project evaluation, especially when comparing projects of different scales and durations.

Let's also consider some details regarding reinvestment rate and MIRR:
- The reinvestment rate $12\% is just one assumption of how positive cash flows will be utilized externally to the project. The MIRR is the calculated overall performance of the entire project considering that 12\% assumption.
- The 18.45% MIRR is a measure of the project's actual profitability over three years (effective annual yield of the project), taking into account the magnitude and timing of the 4000, 5000 and 6000$ inflows. A project with the same reinvestment rate but different cash flows would have a different MIRR.
- The reinvestment rate only addresses what happens to positive cash flows. The MIRR also explicitly includes the cost of financing the initial investment (the 10\% finance rate) in its calculation, providing a more holistic picture of the project's true cost and return.

Here is a Python program to calculate this example:

In [11]:
import numpy_financial as npf

cash_flows = [-10000, 4000, 5000, 6000]
financing_rate = 0.10
reinvestment_rate = 0.12

mirr_value = npf.mirr(cash_flows, financing_rate, reinvestment_rate)

print(f"The project's cash flows are: {cash_flows}")
print(f"Finance Rate: {financing_rate:.2%}")
print(f"Reinvestment Rate: {reinvestment_rate:.2%}")
print(f"The calculated MIRR is: {mirr_value:.4f} or {mirr_value:.2%}")

The project's cash flows are: [-10000, 4000, 5000, 6000]
Finance Rate: 10.00%
Reinvestment Rate: 12.00%
The calculated MIRR is: 0.1845 or 18.45%


Let's consider the same example, but assume that we did not borrow money, this is just our money we're going to invest (let's assume that the finance rate is zero).

In [12]:
import numpy_financial as npf

cash_flows = [-10000, 4000, 5000, 6000]
financing_rate = 0.0
reinvestment_rate = 0.12

mirr_value = npf.mirr(cash_flows, financing_rate, reinvestment_rate)

print(f"The project's cash flows are: {cash_flows}")
print(f"Finance Rate: {financing_rate:.2%}")
print(f"Reinvestment Rate: {reinvestment_rate:.2%}")
print(f"The calculated MIRR is: {mirr_value:.4f} or {mirr_value:.2%}")

The project's cash flows are: [-10000, 4000, 5000, 6000]
Finance Rate: 0.00%
Reinvestment Rate: 12.00%
The calculated MIRR is: 0.1845 or 18.45%


As expected, the result is totally the same as with finance rate 10\% because the outflow is related to the period number 'zero' which is present date, thus the present value of outflow in such a case it's just an outflow itself.

Then, let's assume that we're not going to reinvest inflows. Let's see what happens.

In [13]:
import numpy_financial as npf

cash_flows = [-10000, 4000, 5000, 6000]
financing_rate = 0.0
reinvestment_rate = 0.0

mirr_value = npf.mirr(cash_flows, financing_rate, reinvestment_rate)

print(f"The project's cash flows are: {cash_flows}")
print(f"Finance Rate: {financing_rate:.2%}")
print(f"Reinvestment Rate: {reinvestment_rate:.2%}")
print(f"The calculated MIRR is: {mirr_value:.4f} or {mirr_value:.2%}")

The project's cash flows are: [-10000, 4000, 5000, 6000]
Finance Rate: 0.00%
Reinvestment Rate: 0.00%
The calculated MIRR is: 0.1447 or 14.47%


If we do not reinvest inflows, then
* Total cash inflows = 4000 + 5000 + 6000 = 15000
* Total net profit = 15000 - 10000 = 5000
* Holding Period Return (HPR) is the return on an asset or portfolio over the whole period during which it was held. In our case it is 5000/10000 = 0.5 or 50%. Thus, our investment generated a total return of 50%.
* The 50% is the return over 3 years in our example. To compare it with other annual investments, we can calculate a simple, non-compounded annual return: 50% / 3 years = 16.67 per year.

It is important to note that this is a simple average and does not account for the time value of money (TVM) within the holding period.

**Why MIRR with 0% Reinvestment is Not a Simple Average**

In the last example, we set the reinvestment rate to 0% and found that the `numpy_financial.mirr` function returned **14.47%**. However, a simple calculation of the average annual return gave us a higher figure of **16.67%**.

Let's break down this apparent inconsistency. This is a critical learning point that reveals the core strength of MIRR and other discounted cash flow methods over simple arithmetic averages.

**The Root of the Discrepancy: Time Value of Money (TVM)**

The simple average calculation we performed is as follows:

*   **Total Return (Holding Period Return):** (15,000 - 10,000) / 10,000 = 50%
*   **Simple Average Annual Return:** 50% / 3 years = 16.67% per year

This calculation is misleading because it **ignores the time value of money**. It implicitly assumes that the timing of the cash inflows (4000 in Year 1, 5000 in Year 2, and 6000 in Year 3) does not matter. It treats the entire 15,000 as if it were received at a single point in time.

However, in finance, we know that money received sooner is more valuable than money received later. The MIRR calculation, even with a 0% reinvestment rate, still respects this fundamental principle.

**How MIRR Arrives at 14.47%**

Let's manually re-calculate the MIRR with a 0% reinvestment rate. This will demonstrate precisely how the time value of money is incorporated.

*   **Cash Flows:** `[-10000, 4000, 5000, 6000]`
*   **Finance Rate:** 0%
*   **Reinvestment Rate:** 0%

**Step 1: Present Value of Negative Cash Flows**

The initial investment of 10,000 is already at its present value (t=0).
*   **PV of Outflows = 10000**

**Step 2: Future Value of Positive Cash Flows (Terminal Value)**

With a reinvestment rate of 0%, the cash inflows do not earn any additional return. They are simply held until the end of the project's life.
*   The 4000 from Year 1 is held for 2 years: FV = 4000 * (1 + 0.00)² = 4000
*   The 5000 from Year 2 is held for 1 year: FV = 5000 * (1 + 0.00)¹ = 5000
*   The 6000 from Year 3 is already at the end, so its value is 6000.
*   **Terminal Value** = 4000 + 5000 + 6000 = **15000**

**Step 3: Calculate the MIRR**

The MIRR is the single rate of return that equates the initial investment to the terminal value over the project's life. It answers the question: "At what constant annual rate must my 10000 investment grow to become 15000 in 3 years?"

$$ MIRR = \left( \frac{\text{Terminal Value}}{\text{PV of Outflows}} \right)^{\frac{1}{n}} - 1 $$

$$ MIRR = \left( \frac{15,000}{10,000} \right)^{\frac{1}{3}} - 1 $$

$$ MIRR = (1.5)^{\frac{1}{3}} - 1 $$

$$ MIRR = 1.1447 - 1 = 0.1447 $$

The MIRR is **14.47%**.

**Conclusion: The Meaning of the Two Percentages**

*   **16.67% (Simple Average):** This is the **arithmetic average** return per year. It is a simple, non-compounded rate that incorrectly assumes the total profit of 5000 (15000 / 3 years) was generated evenly over the three years, ignoring the timing of the cash flows.

*   **14.47% (MIRR):** This is the **compounded annual growth rate (CAGR)** of the investment. It is the true, annualized, time-value-of-money-adjusted rate of return. It accurately reflects that the cash inflows are spread out over time. Because not all profits are received immediately, the effective compounded return is lower than the simple average.


### MIRR limitations

The standard MIRR formula has the same limitation as the standard IRR formula: it assumes that all cash flows occur at regular, equally spaced intervals (e.g., annually, monthly, etc.).

The formula for MIRR, which involves raising a value to the power of `1/n`, explicitly uses `n` as the number of *periods*. This structure inherently requires those periods to be of equal length for the calculation to be mathematically sound.

Just as XIRR was developed to handle the irregular timing of cash flows for an internal rate of return calculation, a similar logic is needed for MIRR when dealing with real-world, date-specific cash flows.

While not a standard built-in function in most software like Excel or Google Sheets, financial modelers and analysts have developed a concept often called **XMIRR (Extended Modified Internal Rate of Return)**.

## Extended Modified Internal Rate of Return (XMIRR)

As previously established, the **Modified Internal Rate of Return (MIRR)** represents a significant advancement over the standard IRR by addressing its most critical limitation: the unrealistic assumption that all intermediate cash flows are reinvested at the project's own internal rate of return. MIRR rectifies this by allowing the specification of a more practical and market-reflective reinvestment rate for positive cash flows and a finance rate for negative cash flows.

However, the conventional MIRR, much like the standard IRR, still operates under the implicit assumption that all cash flows occur at regular, equally spaced intervals (e.g., precisely annually or monthly). In the complex reality of financial projects and investments, cash flows rarely adhere to such strict, periodic schedules. Investment contributions, operational expenses, capital calls, and revenue streams often occur on specific, irregular calendar dates.

To bridge this gap and provide an even more accurate and robust measure of project profitability, we use the **Extended Modified Internal Rate of Return (XMIRR)**. XMIRR integrates the date-specific precision of XIRR (which handles irregular cash flow timings) with the realistic reinvestment and finance rate assumptions of MIRR. This combination makes XMIRR a powerful and comprehensive metric for evaluating projects with complex, real-world cash flow patterns.

The **Extended Modified Internal Rate of Return (XMIRR)** is an annualized rate of return that accounts for irregular cash flow dates, a specified finance rate for outflows, and a specified reinvestment rate for inflows. It represents the single, constant growth rate that equates the present value of all project outflows (discounted at the finance rate) to the present value of the terminal value of all project inflows (compounded at the reinvestment rate) over the exact duration of the project.

### Derivation and Formula

The calculation of XMIRR systematically extends the three-step process of MIRR by incorporating the precise dates of each cash flow. This approach is superior because it explicitly separates the treatment of cash outflows (investments or costs) and cash inflows (returns or revenues), using distinct rates that better reflect the firm's actual cost of capital and its opportunities to generate returns.

Let $D_1$ be the earliest date of any cash flow in the series, and $D_n$ be the latest date of any cash flow in the series.
Let $r_f$ be the finance rate, and $r_r$ be the reinvestment rate.

1.  **Calculate the Present Value (PV) of all Negative Cash Flows (Outflows) at the Earliest Date ($D_1$).**
    Each negative cash flow (representing an investment, expense, or capital outflow) is discounted from its actual occurrence date back to the earliest date of the project ($D_1$) using the specified **finance rate ($r_f$)**. The absolute value of these discounted outflows is then summed.

    $$PV_{Outflows} = \sum_{C_i < 0} \frac{|C_i|}{(1 + r_f)^{(D_i - D_1)/365}}$$

    *   **Reasoning:** This step accurately captures the total cost of all initial and subsequent investments (outflows) at a single, consistent reference point (the effective start of the project). The use of the finance rate ($r_f$) realistically reflects the actual cost of funds, which is often the firm's cost of capital, the specific cost of borrowing for the project, or the weighted average cost of capital (WACC). The term $(D_i - D_1)/365$ ensures precise, date-specific discounting, even for fractional years, rather than assuming fixed periodic intervals.

2.  **Calculate the Future Value (FV) of all Positive Cash Flows (Inflows) at the Latest Date ($D_n$).**
    Each positive cash flow (representing a return, revenue, or capital inflow) is compounded forward from its actual occurrence date to the latest date ($D_n$) of the entire project using the specified **reinvestment rate ($r_r$)**. The sum of these compounded values is known as the **Terminal Value (TV)**.

    $$\text{Terminal Value} = \sum_{C_i > 0} C_i \times (1 + r_r)^{(D_n - D_i)/365}$$
    
    *   **Reasoning:** This step realistically aggregates the value of all positive returns (inflows) at the end of the project's life, assuming they are reinvested at a practical, market-determined rate. The reinvestment rate ($r_r$) typically reflects the average rate of return the company expects to earn on alternative investments of similar risk, or its overall cost of capital. The $(D_n - D_i)/365$ term ensures precise, date-specific compounding over fractional years.

3.  **Calculate the XMIRR.**
    The XMIRR is the annualized discount rate that equates the present value of the aggregated outflows (calculated in Step 1) with the present value of the Terminal Value (calculated in Step 2), over the entire duration of the project (from $D_1$ to $D_n$). In essence, it finds the single, constant annualized growth rate that transforms the total initial investment (in PV terms) into the total accumulated returns (in TV terms) over the exact project life.

    The relationship is established as:

    $$PV_{Outflows} = \frac{\text{Terminal Value}}{(1 + XMIRR)^{(D_n - D_1)/365}}$$

    Solving for XMIRR, we get the explicit formula:

    $$XMIRR = \left( \frac{\text{Terminal Value}}{PV_{Outflows}} \right)^{\frac{1}{(D_n - D_1)/365}} - 1$$

    where:
    *   $C_i$: The $i$-th cash flow amount.
    *   $D_i$: The date of the $i$-th cash flow.
    *   $D_1$: The date of the **earliest** cash flow in the series, serving as the common reference point for discounting outflows.
    *   $D_n$: The date of the **latest** cash flow in the series, serving as the common reference point for compounding inflows.
    *   $r_f$: The annual finance rate (used for discounting negative cash flows).
    *   $r_r$: The annual reinvestment rate (used for compounding positive cash flows).
    *   $/365$: Implies the **"Actual/365 Fixed" day count convention**. This means the numerator uses the exact number of calendar days between two dates, while the denominator is a fixed 365 days for annualization, regardless of whether it's a leap year. This convention is standard in many financial functions (like Excel's XIRR/XNPV) for consistency and simplicity.

### Example of XMIRR

To demonstrate the power and precision of XMIRR, let's use the same irregular cash flow series from our XIRR and XNPV examples and apply a more realistic set of financing and reinvestment assumptions.

**Scenario:**
*   **Finance Rate ($r_f$):** 8% (This is the annual rate at which the company borrows funds or its overall cost of capital for financing the project's outflows).
*   **Reinvestment Rate ($r_r$):** 12% (This is the annual rate at which the company can realistically reinvest the positive cash flows generated by the project).

**Cash Flows and Dates:**

| Date          | Cash Flow Amount ($C_i$) |
| :------------ | :----------------------- |
| `2022-01-15`  | `-10000`                 |
| `2022-06-01`  | `-2000`                  |
| `2023-03-10`  | `5000`                   |
| `2024-07-22`  | `10000`                  |
| `2024-12-31`  | `3000`                   |

*   **Earliest Date ($D_1$):** Jan 15, 2022
*   **Latest Date ($D_n$):** Dec 31, 2024

**Step-by-Step Calculation:**

1.  **Calculate the Present Value (PV) of Negative Cash Flows at the Earliest Date ($D_1$).**
    We discount the two outflows back to Jan 15, 2022, using the **8% finance rate**.

    *   $PV(\text{Outflow 1}) = \frac{|-10000|}{(1 + 0.08)^{(Jan 15, 2022 - Jan 15, 2022)/365}} = \frac{10000}{(1.08)^0} = 10,000.00$
    *   $PV(\text{Outflow 2}) = \frac{|-2000|}{(1 + 0.08)^{(Jun 1, 2022 - Jan 15, 2022)/365}} = \frac{2000}{(1.08)^{(137/365)}} = \frac{2000}{1.0293} = 1943.05$
    *   **Total $PV_{Outflows}$ = 10000.00 + 1943.05 = 11943.05**

2.  **Calculate the Future Value (FV) of Positive Cash Flows at the Latest Date ($D_n$).**
    We compound the three inflows forward to Dec 31, 2024, using the **12% reinvestment rate**.

    *   $FV(\text{Inflow 1}) = 5000 \times (1 + 0.12)^{(Dec 31, 2024 - Mar 10, 2023)/365} = 5000 \times (1.12)^{(662/365)} = 5000 \times 1.2282 = 6140.97$
    *   $FV(\text{Inflow 2}) = 10000 \times (1 + 0.12)^{(Dec 31, 2024 - Jul 22, 2024)/365} = 10000 \times (1.12)^{(162/365)} = 10000 \times 1.0516 = 10515.86$
    *   $FV(\text{Inflow 3}) = 3000 \times (1 + 0.12)^{(Dec 31, 2024 - Dec 31, 2024)/365} = 3000 \times (1.12)^0 = 3000.00$
    *   **Terminal Value = 6140.97 + 10515.86 + 3000.00 = 19656.83**

3.  **Calculate the XMIRR.**
    Now, we find the single annualized rate that makes the initial investment ($PV_{Outflows}$) grow to the final accumulated returns (Terminal Value) over the project's total duration.

    *   **Total Duration (in days):** Dec 31, 2024 - Jan 15, 2022 = 1081 days
    *   **Total Duration (in years):** 1081 / 365 = 2.9616 years

    $$XMIRR = \left( \frac{19656.83}{11943.05} \right)^{\frac{1}{2.9616}} - 1$$

    $$XMIRR = (1.6459)^{\frac{1}{2.9616}} - 1$$
    
    $$XMIRR = 1.1832 - 1 = 0.1832$$

The XMIRR is **18.32%**.

This result provides a more nuanced and realistic performance measure than the 21.24% XIRR calculated earlier, as it accounts for the distinct costs of financing and the actual opportunities for reinvesting returns.

In [14]:
from datetime import date

def xmirr(dates, cash_flows, finance_rate, reinvest_rate, day_count_convention=365):
    """
    Calculate Extended Modified Internal Rate of Return (XMIRR)
    for irregular cash flows with dates.
    
    Parameters:
    - dates: list of datetime.date objects
    - cash_flows: list of cash flows (negative for investments, positive for returns)
    - finance_rate: annual financing rate for negative cash flows
    - reinvest_rate: annual reinvestment rate for positive cash flows
    - day_count_convention: number of days in a year for calculations (default 365)
    
    Returns:
    - XMIRR as a decimal (e.g., 0.15 for 15%)
    """
    if len(dates) != len(cash_flows):
        raise ValueError("Dates and cash flows must have the same length")
    
    start_date = min(dates)
    end_date = max(dates)
    total_days = (end_date - start_date).days
    
    if total_days == 0:
        raise ValueError("All dates are the same")
    
    # Calculate present value of negative cash flows (investments)
    pv_negative = 0
    for d, cf in zip(dates, cash_flows):
        if cf < 0:
            years_from_start = (d - start_date).days / day_count_convention
            pv_negative += cf / ((1 + finance_rate) ** years_from_start)
    
    # Calculate future value of positive cash flows (returns)
    fv_positive = 0
    for d, cf in zip(dates, cash_flows):
        if cf > 0:
            years_to_end = (end_date - d).days / day_count_convention
            fv_positive += cf * ((1 + reinvest_rate) ** years_to_end)
    
    if pv_negative == 0:
        raise ValueError("No negative cash flows found")
    if fv_positive == 0:
        raise ValueError("No positive cash flows found")
    
    # Calculate XMIRR
    years_total = total_days / day_count_convention
    xmirr = (fv_positive / abs(pv_negative)) ** (1 / years_total) - 1
    
    return xmirr


# Your example
dates = [
    date(2022, 1, 15),
    date(2022, 6, 1),
    date(2023, 3, 10),
    date(2024, 7, 22),
    date(2024, 12, 31)
]

cash_flows = [-10000, -2000, 5000, 10000, 3000]

financing_rate = 0.08
reinvestment_rate = 0.12

xmirr_value = xmirr(dates, cash_flows, financing_rate, reinvestment_rate)

print("Date\t\t\tCash Flow")
print("-" * 40)
for d, cf in zip(dates, cash_flows):
    print(f"{d}\t\t{cf:>8}")
print("-" * 40)
print(f"\nFinancing Rate: {financing_rate:.2%}")
print(f"Reinvestment Rate: {reinvestment_rate:.2%}")
print(f"The calculated XMIRR is: {xmirr_value:.4f}")
print(f"As a percentage: {xmirr_value:.2%}")

Date			Cash Flow
----------------------------------------
2022-01-15		  -10000
2022-06-01		   -2000
2023-03-10		    5000
2024-07-22		   10000
2024-12-31		    3000
----------------------------------------

Financing Rate: 8.00%
Reinvestment Rate: 12.00%
The calculated XMIRR is: 0.1832
As a percentage: 18.32%


### Advantages and Interpretation of XMIRR

XMIRR offers significant advantages over simpler metrics, making it particularly valuable for sophisticated capital budgeting and investment analysis in complex scenarios:

1.  **Realistic Reinvestment and Financing Assumptions:** This is the primary strength of XMIRR. By allowing distinct, user-defined reinvestment and finance rates, XMIRR directly addresses the unrealistic implicit reinvestment assumption of IRR/XIRR. It provides an economically plausible and defensible measure of return that reflects the firm's actual cost of capital and its opportunities to reinvest positive cash flows at prevailing market rates or its average cost of capital.
2.  **Accurate Handling of Irregular Cash Flows:** XMIRR precisely accounts for the exact timing of cash flows using fractional year exponents based on actual calendar dates. This makes it robust for real-world projects where investments and returns do not occur at fixed, periodic intervals, providing a more accurate reflection of the economic exposure over time.
3.  **Unique and Unambiguous Solution:** Unlike IRR/XIRR, which can lead to multiple real solutions (or no real solution) when dealing with "non-conventional" cash flow patterns (i.e., cash flows that switch signs more than once), XMIRR generally provides a unique and unambiguous solution. This is because it transforms the problem into equating a single present value of outflows with a single future value (terminal value) of inflows, thereby avoiding the complex polynomial root-finding problem inherent in traditional IRR calculations.
4.  **Strong Consistency with NPV:** XMIRR exhibits strong theoretical consistency with the Net Present Value (NPV) criterion, especially when both the finance rate ($r_f$) and the reinvestment rate ($r_r$) are set equal to the project's cost of capital (which is typically the discount rate 'r' used for NPV). Under these conditions, XMIRR will generally lead to the same accept/reject decision as NPV for independent projects. Furthermore, it tends to be more consistent with NPV than IRR when evaluating mutually exclusive projects, which is critical because NPV is theoretically superior for maximizing shareholder wealth.
5.  **Comprehensive Economic View:** By integrating both the cost of financing (for outflows) and the expected return from reinvestment (for inflows), XMIRR offers a holistic and economically sound perspective on a project's profitability over its entire irregular timeline. This makes it a powerful tool for managers making critical investment decisions.
6.  **Annualized Rate for Comparison:** Like XIRR, XMIRR always returns an annualized rate, facilitating direct comparison with other annualized investment returns (e.g., bond yields, stock market returns, or a company's WACC).

---

## Comparison of Metrics

To consolidate the concepts discussed, the following table provides a comparative overview of all the advanced rate of return metrics. This summary helps in selecting the most appropriate tool for a given financial analysis scenario.

| Metric | Measures...                                                              | Key Assumption for Intermediate Cash Flows                                       | Handles Irregular Timings? | Best Used For...                                                                                                                              |
| :--- | :----------------------------------------------------------------------- | :------------------------------------------------------------------------------- | :--- | :-------------------------------------------------------------------------------------------------------------------------------------------- |
| **NPV** | The absolute **dollar value** a project adds to the firm (Present Value of Inflows - PV of Outflows). | Reinvested at the **discount rate** (cost of capital). This is generally realistic. | No   | The gold standard for capital budgeting, especially when comparing mutually exclusive projects of different scales or durations.          |
| **XNPV** | The absolute **dollar value** a project adds, accounting for the **exact dates** of cash flows. | Reinvested at the **discount rate**.                                               | Yes  | Precise valuation of projects where cash flows do not occur at regular yearly or monthly intervals. The most accurate valuation tool. |
| **IRR** | The **percentage rate of return** at which the project's NPV equals zero.                            | Reinvested at the **calculated IRR itself**. This can be highly unrealistic.        | No   | Quickly assessing the potential return of a simple project with regular cash flows. Good for communicating returns as a percentage.     |
| **XIRR** | The **annualized percentage rate of return** for a series of cash flows occurring on **irregular dates**. | Reinvested at the **calculated XIRR itself**. Still an unrealistic assumption.   | Yes  | Accurately measuring the annualized performance of investments with irregular contributions and withdrawals (e.g., personal portfolios). |
| **MIRR** | A modified **percentage rate of return** that addresses IRR's reinvestment flaw.                     | Reinvested at a **user-defined rate** (e.g., cost of capital). This is realistic. | No   | Providing a more reliable and theoretically sound rate of return than IRR for projects with regular cash flows.                         |
| **XMIRR** | The most robust **annualized percentage rate of return**, combining date precision and realistic reinvestment. | Reinvested at a **user-defined rate** and financed at a **user-defined rate**. | Yes  | The most comprehensive analysis for complex, real-world projects with irregular cash flows and varying costs of capital/reinvestment opportunities. |

## Additional materials

* https://en.wikipedia.org/wiki/Cost_of_capital
* https://en.wikipedia.org/wiki/Weighted_average_cost_of_capital
* https://en.wikipedia.org/wiki/Time_value_of_money
* https://en.wikipedia.org/wiki/Present_value
* https://en.wikipedia.org/wiki/Future_value
* https://en.wikipedia.org/wiki/Minimum_acceptable_rate_of_return
* https://en.wikipedia.org/wiki/Holding_period_return
* https://en.wikipedia.org/wiki/Net_present_value
* https://en.wikipedia.org/wiki/Internal_rate_of_return
* https://en.wikipedia.org/wiki/Discounted_cash_flow
* https://en.wikipedia.org/wiki/Modified_internal_rate_of_return
* https://en.wikipedia.org/wiki/Capital_budgeting
* https://en.wikipedia.org/wiki/Discounting
* https://en.wikipedia.org/wiki/Compound_interest
* Youtube: Modified Internal Rate of Return (MIRR) - Basics, Formula, Calculations in Excel (Step by Step) https://www.youtube.com/watch?v=lCaOyFJC_Po